tools/release/adot-operator-images-mirror/mirror.go

```diff
@@ -0,0 +1,244 @@
+			continue
+		}
+
+		if err := m.pushImage(tag.Name); err != nil {
+			log.Printf("Failed to push retagged image %s:%s: %v", m.targetRepositoryName(), tag.Name, err)
+			continue
+		}
+	}
+}
+
+func (m *mirror) getRemoteTags() ([]RepositoryTag, error) {
+	var url string
+	switch m.sourceRepo.Host {
+	case quay:
+		url = fmt.Sprintf("https://quay.io/api/v1/repository/%s/tag", m.sourceRepositoryName())
+	case gcr:
+		url = fmt.Sprintf("https://gcr.io/v2/%s/tags/list", m.sourceRepositoryName())
+	}
+
+	var (
+		allTags []RepositoryTag
+		err     error
+		res     *http.Response
+		req     *http.Request
+		retries = 5
+	)
+
+	// Try to fetch the tags information through HTTP get request.
+	for retries > 0 {
+		req, err = http.NewRequest("GET", url, nil)
+		if err != nil {
+				return nil, err
+			}
+
+		res, err = httpClient.Do(req)
+		if err != nil {
+			log.Printf("Failed to get %s, retrying", url)
+			retries--
+		} else if res.StatusCode == 429 {
+			sleepTime := getSleepTime(res.Header.Get("X-RateLimit-Reset"), time.Now())
+			log.Printf("Rate limited on %s, sleeping for %s", url, sleepTime)
+			time.Sleep(sleepTime)
+			retries--
+		} else if res.StatusCode < 200 || res.StatusCode >= 300 {
+			log.Printf("Get %s failed with %d, retrying", url, res.StatusCode)
+			retries--
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	dc := json.NewDecoder(res.Body)
+
+	// Decode the response and add the tags to remoteTags field of mirror struct.
+	switch m.sourceRepo.Host {
+	case quay:
+		var tags QuayTagsResponse
+		if err := dc.Decode(&tags); err != nil {
+			return nil, err
+		}
+		allTags = append(allTags, tags.Tags...)
+	case gcr:
+		var tags GCRTagsResponse
+		if err := dc.Decode(&tags); err != nil {
+			return nil, err
+		}
+		for _, tag := range tags.Tags {
+			allTags = append(allTags, RepositoryTag{
+				Name: tag,
+			})
+		}
+	}
+
+	log.Printf("Finished scraping remote tags from %s", m.sourceRepositoryFullName())
+
+	return allTags, nil
+}
+
+func (m *mirror) filterTags() {
+	tmp := make([]RepositoryTag, 0)
+
+	for _, remoteTag := range m.remoteTags {
+		// We only keep the tags starting with "v" or with the name "latest".
+		if string(remoteTag.Name[0]) == "v" || remoteTag.Name == "latest" {
+			tmp = append(tmp, remoteTag)
+		}
+	}
+
+	m.remoteTags = tmp
+}
+
+func (m *mirror) sourceRepositoryName() string {
+	return fmt.Sprintf("%s/%s", m.sourceRepo.Registry, m.sourceRepo.Name)
+}
+
+func (m *mirror) sourceRepositoryFullName() string {
+	return fmt.Sprintf("%s/%s/%s", m.sourceRepo.Host, m.sourceRepo.Registry, m.sourceRepo.Name)
+}
+
+func (m *mirror) targetRepositoryName() string {
+	return fmt.Sprintf("%s/%s", m.targetRepo.Registry, m.targetRepo.Name)
+}
+
+func getSleepTime(rateLimitReset string, now time.Time) time.Duration {
+	rateLimitResetInt, err := strconv.ParseInt(rateLimitReset, 10, 64)
```

Comment:
not sure if this constant sleep time is longer enough for rate throttling reset, and exponential backoff retries seems better :)

Reply:
Sure, I will use exponential backoff retries for this. Thanks!

Reply:
Hi Min, I refactored the code structure in mirror.go to apply exponential backoff algorithm to the HTTP request as well as keeping the original sleep time if we get the `X-RateLimit-Reset` value from HTTP response header.